### PR TITLE
Consolidate all package configuration using Poetry

### PR DESCRIPTION
**Pull request by BrunoGrandePhD**
**N.B.** This PR builds on top of PR #387, so you can opt to review this one instead of #387. 

My original goal here was to create an isolated development environment for `schematic`. I first looked into `pipenv`, but I soon found many recommendations for [`poetry`](https://python-poetry.org/). Briefly, `poetry` is a one-stop shop for most Python development tasks, including dependency management (in lieu of `pip`), environment management (in lieu of `virtualenv` or `conda`), and deployment (in lieu of `setuptools`). 

I was able to consolidate several configuration files (_i.e._ `requirements.txt`, `test-requirements.txt`, `setup.py`, `pytest.ini`) into the `pyproject.toml` `poetry` config. I've been able to perform a [test deployment](https://pypi.org/project/schematic-test/) to PyPI as a proof of concept. Unfortunately, the `schematic` name is already taken on PyPI, so we'll need to update `pyproject.toml` accordingly. 

I also updated the GitHub CI to use `poetry` to install dependencies instead of `pip`. I also added steps to cache dependencies for faster testing. 

Here are a few `poetry` commands that are useful to know: 

```bash
# Track a new package and version
poetry add <package>

# Create new environment and install dependencies
poetry install

# Build and deploy to PyPI
poetry deploy --build

# Run commands from within environment
poetry run <command>
poetry run pytest
poetry run schematic [...]
```

### CODE REVIEW

**Review by milen-sage — COMMENTED** [review]
awesome!

let's fix the pyyaml and merge.

**Comment by BrunoGrandePhD** [comment]
I've updated the PyProject TOML file, including the name, authors, keywords, and list of dependencies (restricted to those appearing in an `import` statement).

File: pyproject.toml [hunk 1]
```diff
@@ -51,19 +45,18 @@ importlib-metadata = {version = "^1.6.0", python = "<3.8"}
 inflection = "^0.5.1"
 jsonschema = "^3.2.0"
 networkx = "^2.5"
+oauth2client = "<4.0.0"  # Specified because of bug in version ^4.0.0
```
Comment (milen-sage) [inline pyproject.toml:48]: yes, we need to evaluate our codebase and specify what would need to be done to to transition to later versions of this lib; but that's different PR :)

**Review by milen-sage — APPROVED** [review]

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    "Xengie Doan <xengie.doan@sagebase.org>"
+]
+
+readme = "README.md"
+
+homepage = "https://github.com/Sage-Bionetworks/schematic"
+repository = "https://github.com/Sage-Bionetworks/schematic"
+documentation = "https://github.com/Sage-Bionetworks/schematic"
+
+keywords = ["schema", "metadata", "validation", "data model", "linked data"]
```
Comment (sujaypatil96) [inline pyproject.toml:18]: Just to note a few more keywords (in addition to the current list):
- schema
- data model
- linked data
- metadata
- validation
- annotation
- knowledge graph/ontology
- ingress

No need to add them (_you can if you want_) in this PR, just noting them for reference.

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+google-api-python-client = "^1.12.8"
+google-auth-httplib2 = "^0.0.4"
+google-auth-oauthlib = "^0.4.2"
+graphviz = "^0.16"
+importlib-metadata = {version = "^1.6.0", python = "<3.8"}
+inflection = "^0.5.1"
+jsonschema = "^3.2.0"
+networkx = "^2.5"
+oauth2client = "<4.0.0"  # Specified because of bug in version ^4.0.0
+orderedset = "^2.0.3"
```
Comment (sujaypatil96) [inline pyproject.toml:49]: Feel free to remove this unused import from schema generator: [`from orderedset import OrderedSet`](https://github.com/Sage-Bionetworks/schematic/blob/develop/schematic/schemas/generator.py#L4), and remove the dependency from the `.toml` file as well.

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+pandas = "^1.2.1"
+pygsheets = "^2.0.4"
+PyYAML = "^5.4.1"
+rdflib = "^5.0.0"
+setuptools = "^52.0.0"
+synapseclient = "^2.2.2"
+tabletext = "^0.1"
```
Comment (sujaypatil96) [inline pyproject.toml:56]: You can remove this import statement from here: [`import tabletext`](https://github.com/Sage-Bionetworks/schematic/blob/develop/schematic/schemas/explorer.py#L7), and also this print statement from here: [`print(tabletext.to_text(content))`](https://github.com/Sage-Bionetworks/schematic/blob/develop/schematic/schemas/explorer.py#L263). I suspect we might do something different for logging purposes, so we can add a different logging statement when dealing with issue #317.

**Review by sujaypatil96 — APPROVED** [review]
Mentioned a few nits in the review. It's not critical that they be incorporated, but would be nice if they were. Approving this PR, so feel free to make any changes in a new commit and merge the PR.🚀

File: pyproject.toml (hunk 2 above)
Comment (BrunoGrandePhD) [reply pyproject.toml:18]: I'm limited to 5 keywords ([source](https://python-poetry.org/docs/pyproject/#keywords)), so I added the additional ones as a comment for future reference.

File: pyproject.toml (hunk 3 above)
Comment (BrunoGrandePhD) [reply pyproject.toml:49]: Good catch!

File: pyproject.toml (hunk 4 above)
Comment (BrunoGrandePhD) [reply pyproject.toml:56]: Thanks!

**Comment by BrunoGrandePhD** [comment]
🚀

**Pull request merged**